From: Chandan Rajendra <chandan@linux.vnet.ibm.com>
Date: Sun, 25 Dec 2016 19:01:03 +0530
Subject: clean_bdev_aliases: Prevent cleaning blocks that are not in block
 range
References: bsc#1020989,FATE#322379
Git-commit: 6c006a9d94bfb5cbcc5150e8fd7f45d3f92f3ee8
Patch-Mainline: v4.10-rc3

The first block to be cleaned may start at a non-zero page offset. In
such a scenario clean_bdev_aliases() will end up cleaning blocks that
do not fall in the range of blocks to be cleaned. This commit fixes the
issue by skipping blocks that do not fall in valid block range.

Signed-off-by: Chandan Rajendra <chandan@linux.vnet.ibm.com>
Reviewed-by: Jan Kara <jack@suse.cz>
Reviewed-by: Eryu Guan <eguan@redhat.com>
Signed-off-by: Jens Axboe <axboe@fb.com>
Signed-off-by: Hannes Reinecke <hare@suse.de>
---
 fs/buffer.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/fs/buffer.c b/fs/buffer.c
index e829ee1..bdb5346 100644
--- a/fs/buffer.c
+++ b/fs/buffer.c
@@ -1671,7 +1671,7 @@ void clean_bdev_aliases(struct block_device *bdev, sector_t block, sector_t len)
 			head = page_buffers(page);
 			bh = head;
 			do {
-				if (!buffer_mapped(bh))
+				if (!buffer_mapped(bh) || (bh->b_blocknr < block))
 					goto next;
 				if (bh->b_blocknr >= block + len)
 					break;
-- 
1.8.5.6

